pages/products-page.ts

```diff
@@ -0,0 +1,199 @@
+            const productToRemove = this.cartProductWrappers.filter({hasText: product.name});
+            await productToRemove.getByRole('link', { name: 'Remove' }).click();
+            await this.page.getByRole('dialog').getByRole('button', {name: 'OK'}).click();
+            await productToRemove.waitFor({state: "detached"});
+        })
+    }
+
+    /**
+     * @param doQuickCheck quick check reads product count from badge near shopping cart icon,
+     * if false, check count by opening shopping cart and summing products
+     * @returns 
+     */
+    @boxedStep
+    public async getProductsInCartCount(doQuickCheck: boolean = true): Promise<number> {
+        if(doQuickCheck) {
+            return await this.cartWrapper.locator('.counter-number').textContent({timeout: 1000}).then( numberAsText => Number.parseInt(numberAsText ?? '0'));
+        } else {
+            return await this.withShoppingCartOpen( async () => {
+                return await this.cartProductWrappers.count();
+            })
+        }
+    }
+
+    /**
+     * @param doQuickCheck quick check reads product count from badge near shopping cart icon,
+     * if false, check emptiness by opening shopping cart and looking for products
+     * @returns 
+     */
+    @boxedStep
+    public async isCartEmpty(doQuickCheck: boolean = true): Promise<boolean> {
+        return await this.getProductsInCartCount(doQuickCheck).then(count => count === 0);
+    }
+
+    /**
+     * Get array of names of items in cart
+     * Preserves cart state (if it was open before call, it will be open after)
+     */
+    @boxedStep
+    public async getCartContents(): Promise<Set<string>> {
+        // TODO: change return type into Product
+        // TODO: do I need to open cart for this
+        return await this.withShoppingCartOpen( async () => {
+            const productsInCart: Set<string> = new Set();
+            await this.cartProductWrappers.first().waitFor({state: 'visible'});
+            for( const cartProductWrapper of await this.cartProductWrappers.all() ) {
+                productsInCart.add(await this.getProductFromCartItem(cartProductWrapper));
+            }
+            return productsInCart;
+        })
+    }
+
+    /**
+     * execute action with contents of shopping cart while preserving it's state
+     * (if it's open it will remain open, if closed, it will be closed)
+     * @param action action to execute with open shopping cart
+     * @returns 
+     */
+    private async withShoppingCartOpen<T>( action: () => T ):  Promise<T> {
+        const cardOpenedBefore = await this.isCartOpen();
+        if(!cardOpenedBefore) {
+            await this.openCart();
+        }
+
+        const result: T = await action();
+
+        if(!cardOpenedBefore) {
+            await this.closeCart();
+        }
+
+        return result;
+    }
+    
+    @boxedStep
+    private async getProductFromCartItem(cartProductWrapper: Locator): Promise<string> {
```

Comment:
For these purposes you need to create Component, something like that:
```
export default class BaseComponent {
    readonly container: string;

    constructor(container: string) {
        this.container = container;
    }
}
```


```
export class Product extends BaseComponent {

    constructor(container: string) {
        super(container);
    }
```
}